### PR TITLE
[mobile-client-build-tab] add mobile client build tab to the MAR view

### PR DESCRIPTION
**Pull request by JameelB**
## Description
Changes includes:
- Adding build tab
- Displaying build pipelines within that build tab
- Displaying build config, latest builds and build history of each pipeline
- Displaying download link and qr code

## Verification Steps
**Setup**
- Provision a mobile client
- Verify that the `Builds` tab is not available within the mobile client view as there is no mobile ci/cd provisioned
- Provision a mobile ci/cd service
- Verify that the `Builds` tab is available within the mobile client view.
- Verify that a message shows within the `Builds` tab stating `No Build Config`

**Creating Builds**
- Create a build
- Verify that a new build is created and shown within the build tab
- Verify that there is a `Start Build` button available in the header
- Expand the build to see the build overview
- Verify that a `Build Config` section is available
- Verify that the contents of the `Build Config` is correct
- Verify that a `Builds` section is available
- Verify that a message stating `No Builds` is shown and a `Start Build` button is displayed within the `Builds` section.
- Start the build
- Verify that a success notification appears
- Verify that a build process has started and the latest build is displayed within the `Builds` section with the pipeline stages.
- Verify that the status icon within the build view and the header corresponds correctly to the build process status
- Verify that the links direct to the correct places.
- Verify that a download button appears only when the build completes successfuly
- Click the download button
- Verify that the URL is correct and a qr code is displayed

**Build History**
- Verify that `Show build history` is available
- Expand the build history
- Verify that the previous builds are shown and a download button is only available for builds that completed successfully
- Verify that when the `Download` button is clicked, the URL and qr code for the artefact is shown for that build

**Cleanup**
- Verify that a build can be deleted within the `Builds` tab

## Additional Notes
Related JIRAS:
- https://issues.jboss.org/browse/AEROGEAR-2590
- https://issues.jboss.org/browse/AEROGEAR-2592
- https://issues.jboss.org/browse/AEROGEAR-2593
- https://issues.jboss.org/browse/AEROGEAR-2888

### CODE REVIEW

**Comment by JameelB** [comment]
@sedroche @maleck13 @dimitraz Atm, the build tab does not show when there is no mobile ci/cd provision. Should this be displayed at all times and if there is no mobile ci/cd provisioned, we can show a message stating `no mobile ci/cd provisioned` and allow to provision a mobile ci/cd service from this view?

File: app/views/mobile-client-download-url.html [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<div ng-if="$ctrl.showPanel" class="mobile-client-download-url col-md-12">
+  <div ng-if="!($ctrl.mobileBuild | annotation : $ctrl.BUILD_DOWNLOAD_ANNOTATION)" class="spinner spinner-lg" aria-hidden="true"></div>
+  <div ng-if="$ctrl.mobileBuild | annotation : $ctrl.BUILD_DOWNLOAD_ANNOTATION">
+    <div class="col-md-6">
+      <p>Download from URL:</p>
+      <a ng-href="{{$ctrl.mobileBuild | annotation : $ctrl.BUILD_DOWNLOAD_ANNOTATION}}">{{$ctrl.mobileBuild | annotation : $ctrl.BUILD_DOWNLOAD_ANNOTATION}}</a>
```
Comment (dimitraz) [inline app/views/mobile-client-download-url.html]: Can the target be `_blank` here

File: app/scripts/directives/mobileClientBuildsList.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    $location,
+    APIService,
+    DataService
+  ) {
+    var ctrl = this;
+    var buildConfigsVersion = APIService.getPreferredVersion('buildconfigs');
+    var watches = [];  
+
+    ctrl.$onChanges = function(changes) {
+      var mobileClientChanges = changes.mobileClient && changes.mobileClient.currentValue;
```
Comment (philbrookes) [inline app/scripts/directives/mobileClientBuildsList.js]: would this be easier to read as:
```
var mobileClientChanged = _.get(changes, "mobileClient.currentValue", false);
```

File: app/scripts/directives/mobileClientBuildsRow.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    row.state = {};
+    row.buildConfigsInstantiateVersion = APIService.getPreferredVersion('buildconfigs/instantiate');
+    row.buildConfigsVersion = APIService.getPreferredVersion('buildconfigs');
+    var buildsVersion = APIService.getPreferredVersion('builds');
+    var watches = [];
+
+    _.extend(row, ListRowUtils.ui);
+
+    row.$onChanges = function(changes) {
+      var apiObjectChanges = changes.apiObject && changes.apiObject.currentValue;
```
Comment (philbrookes) [inline app/scripts/directives/mobileClientBuildsRow.js]: might be nicer to use the _.get here too, what do you think?

File: app/scripts/directives/mobileClientDownloadUrl.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    ctrl.$onChanges = function(changes) {
+      var mobileBuildChanges = changes.mobileBuild && changes.mobileBuild.currentValue;
+      if (mobileBuildChanges && !ctrl.context) {
+        ctrl.context = {namespace:_.get(ctrl, 'mobileBuild.metadata.namespace')};
+      }
+
+      var showPanelChanges = changes.showPanel && changes.showPanel.currentValue;
+      if (showPanelChanges) {
+        var downloadUrl = ctrl.mobileBuild.metadata.annotations[ctrl.BUILD_DOWNLOAD_ANNOTATION];
```
Comment (philbrookes) [inline app/scripts/directives/mobileClientDownloadUrl.js:40]: I'm not sure this belongs in onChanges, if you load the page with a build that already has a download link generated, it should show the QR code immediately right? This looks like it won't show it until the build is updated somehow.

Comment (philbrookes) [reply app/scripts/directives/mobileClientDownloadUrl.js:40]: It's worth testing that anyway, and if it doesn't work, putting this logic inside a watch might work better?

Comment (JameelB) [reply app/scripts/directives/mobileClientDownloadUrl.js:40]: if the qr code/download link is already generated, it should be available in the annotations which it checks first. So even if you reload the page, the download url should already be available.

File: app/scripts/directives/mobileClientBuildsRow.js (hunk 3 above)
Comment (sedroche) [reply app/scripts/directives/mobileClientBuildsRow.js]: :+1:

File: app/scripts/directives/mobileClientDownloadUrl.js (hunk 4 above)
Comment (philbrookes) [reply app/scripts/directives/mobileClientDownloadUrl.js:40]: yeah, correct, this is fine. I have tested it locally.

**Review by philbrookes — APPROVED** [review]

**Review by dimitraz — APPROVED** [review]
verified locally :+1:

**Pull request merged**